### PR TITLE
[ZEPPELIN-2403] interpreter property widgets

### PR DESCRIPTION
**Pull request by tinkoff-dwh**
### What is this PR for?
I spoiled the previous PR #2251

Added widgets (string, text, url, password, url, checkbox) to properties of interpreters. Those are widgets for properties customization. Properties must have the ability to customize the display (for example password).

### What type of PR is it?
Feature

### What is the Jira issue?
https://issues.apache.org/jira/browse/ZEPPELIN-2403

### How should this be tested?
- remove conf/interpreter.json
- Try new form (create, edit) of interpreter settings

### Screenshots (if appropriate)
edit
edit

view
view

### Questions:
* Does the licenses files need update? no
* Is there breaking changes for older versions? no
* Does this needs documentation? no

### CODE REVIEW

**Comment by 1ambda** [comment]
1. number, string and url can be generalized as `text`. For url, we can use `ng-html` with angular's `$sse` service to draw `<a>..</a>` element
2. if `text` widget means `textarea` DOM, can we change it to `textarea` instead of?

**Comment by tinkoff-dwh** [comment]
@1ambda 
I think it's too complicate code, for example generate list (combobox) of widgets. How will the selection of the type of the password, second combobox?. I see no reason to generalize number, string and url.

**Comment by 1ambda** [comment]
One example of separating widget with value types.

```
// https://github.com/apache/zeppelin/blob/master/zeppelin-web/src/app/tabledata/advanced-transformation-util.test.js#L19-L25

const MockParameter = {
  'floatParam': { valueType: 'float', defaultValue: 10, description: '', },
  'intParam': { valueType: 'int', defaultValue: 50, description: '', },
  'jsonParam': { valueType: 'JSON', defaultValue: '', description: '', widget: 'textarea', },
  'stringParam1': { valueType: 'string', defaultValue: '', description: '', },
  'stringParam2': { valueType: 'string', defaultValue: '', description: '', widget: 'input', },
  'boolParam': { valueType: 'boolean', defaultValue: false, description: '', widget: 'checkbox', },
  'optionParam': { valueType: 'string', defaultValue: 'line', description: '', widget: 'option', optionValues: [ 'line', 'smoothedLine', ], },
}
```

1. In case of password, value type will be decided by its definition. If it's number, we can validate it as number. (string as well)

2. In case of combobox, type will be boolean. 

3. We can parse values based on their type like https://github.com/apache/zeppelin/blob/master/zeppelin-web/src/app/tabledata/advanced-transformation-util.js#L85-#L117

4. I think it's not complicated. IMO, it's the part we should design carefully since it's hard to change (backward compatibility)

**Comment by tinkoff-dwh** [comment]
CI red (checkstyle errors in all modules)

**Comment by tinkoff-dwh** [comment]
Ready to review

**Comment by 1ambda** [comment]
Let me review and comment soon for this awesome improvement.

**Comment by 1ambda** [comment]
Sorry for the late review. 

seems that it doesn't work with existing settings. Even can't start Zeppelin instance.

```
Caused by: java.lang.ClassCastException: java.lang.String cannot be cast to com.google.gson.internal.StringMap
        at org.apache.zeppelin.interpreter.InterpreterSettingManager.loadFromFile(InterpreterSettingManager.java:174)
        at org.apache.zeppelin.interpreter.InterpreterSettingManager.init(InterpreterSettingManager.java:345)
        at org.apache.zeppelin.interpreter.InterpreterSettingManager.<init>(InterpreterSettingManager.java:150)
        at org.apache.zeppelin.server.ZeppelinServer.<init>(ZeppelinServer.java:149)
        ... 29 more
```

- It would be nicer to provide migration function like https://github.com/apache/zeppelin/blob/master/zeppelin-zengine/src/main/java/org/apache/zeppelin/notebook/Notebook.java#L418
- or use the default type / widget if a field doens't have (e.g `string`, `input`)

File: docs/rest-api/rest-interpreter.md [hunk 1]
```diff
@@ -302,7 +327,11 @@ The role of registered interpreters, settings and interpreters group are describ
     "name": "Markdown setting name",
     "group": "md",
     "properties": {
-      "propname": "propvalue"
+      "propname": {
+        "name": "propname",
```
Comment (1ambda) [inline docs/rest-api/rest-interpreter.md]: Do we need the `name` field?

**Comment by tinkoff-dwh** [comment]
@1ambda 
i wrote about migration in upgrade.md. I'll see Notebook.java if the documentation is not enough

File: docs/rest-api/rest-interpreter.md (hunk 1 above)
Comment (tinkoff-dwh) [inline docs/rest-api/rest-interpreter.md]: i think not, but interpreter-settings.json have this field...

**Comment by 1ambda** [comment]
As far as I Know, Zeppelin supports backward compatibility even for interpreter settings.

- I think we need some migration code w/ documentation

**Comment by Leemoonsoo** [comment]
This PR adds two new field 'widget' and 'type' to each properties.

'widget' : input, checkbox, textarea, password
'type' : number, string, boolean, url, password

How about just add single field 'type' with possible values "string"(input), "number", "url", "password", "checkbox", "text"(textarea) ?

@1ambda mentioned about backward compatibility. It'll help users who upgrade Zeppelin, if this PR reads `conf/interpreter.json` generated by previous version.

**Comment by tinkoff-dwh** [comment]
@Leemoonsoo 
the separation discussed above in the comments

**Comment by tinkoff-dwh** [comment]
CI red https://travis-ci.org/tinkoff-dwh/zeppelin/builds/231079773

another tests 
  SecurityRestApiTest.testGetUserList:69 » JsonSyntax java.lang.IllegalStateExce...

  SecurityRestApiTest.testTicket:55 » JsonSyntax java.lang.IllegalStateException...

**Comment by tinkoff-dwh** [comment]
ready to review

**Comment by 1ambda** [comment]
LGTM

**Comment by Leemoonsoo** [comment]
> the separation discussed above in the comments

maybe i missed, @tinkoff-dwh could you point out the comment?

**Comment by tinkoff-dwh** [comment]
@Leemoonsoo 
first three comments

...
I think it's not complicated. IMO, it's the part we should design carefully since it's hard to change (backward compatibility)
...
widget means UI, and type defines their actual value types. Sometime they might be mixed together, but usually not.

**Comment by Leemoonsoo** [comment]
If i list all combinations of widget-type in this PR,

| widget  | type |
| ------------- | ------------- |
| input  | **string**  |
| input | **number**  |
| **textarea** | string |
| **password** | password |
| **checkbox** | boolean |
| input | **url** |

As we can see, each combination can be uniquely identified by either widget or type. (bold font)
Unless there's big plan in the future add a lot of combinations of those, i think just keeping one field 'type' is enough to identify all possible combinations.

**Comment by tinkoff-dwh** [comment]
looks good. initially, the idea with the type well, as the type indicated the type of the stored value (widget is pointing to), but **password** all broke) 

@1ambda 
what do you think?

**Comment by 1ambda** [comment]
IMO, It's ok if we are not going to add other types like `JSON` which can be `type` and used for other widgets (e.g `textarea`, ...)

**Comment by tinkoff-dwh** [comment]
@Leemoonsoo 
done

**Comment by tinkoff-dwh** [comment]
@Leemoonsoo 
?

**Comment by tinkoff-dwh** [comment]
Ready to review

**Comment by Leemoonsoo** [comment]
@tinkoff-dwh Let me take a look!

**Comment by tinkoff-dwh** [comment]
Ready to review :)

**Comment by tinkoff-dwh** [comment]
Ready to review

**Comment by tinkoff-dwh** [comment]
@Leemoonsoo @1ambda 
maybe merge? 2 month :)

**Comment by 1ambda** [comment]
I am not a big fan of this approach.  But reviewers discussed and I partially agree in terms of complexity. 

I will check few things since it was quite old when I last reviewed this, 
Then if it's good will merge if no more discussion.

File: docs/usage/rest_api/interpreter.md [hunk 2]
```diff
@@ -153,8 +156,16 @@ The role of registered interpreters, settings and interpreters group are describ
...
       "properties": {
-        "spark.cores.max": "",
-        "spark.executor.memory": "1g",
+        "spark.cores.max": {
+          "name": "",
+          "value": "spark.cores.max",
+          "type": "number"
+        },
+        "spark.executor.memory": {
+          "name": "",
```
Comment (1ambda) [inline docs/usage/rest_api/interpreter.md]: is it ok to have name `""`? 

other settings have 

image

**Comment by 1ambda** [comment]
1. Please resolve the conflict.
2. Left a minor comment please check that.
3. Tested locally and works well with 6 newly added types. 
4. Backward compatibility will be handed by the `convertFlatPropertiesToPropertiesWithWidgets` func (check interpreter config and convert old format to new format)

After 1, 2 are resolved, merge it if no more discussion.

**Comment by tinkoff-dwh** [comment]
1,2 done
@1ambda thx for review

**Comment by 1ambda** [comment]
LGTM, merge if no more discussion.

**Comment by 1ambda** [comment]
@tinkoff-dwh Please resolve conflict.

Usually, we merge after 1 day when someone left LGTM, but this time, conflict occurred before 1 day.

**Comment by tinkoff-dwh** [comment]
@1ambda 
done

**Comment by 1ambda** [comment]
merge if no more discussion.

**Pull request closed without merging**